File: libs/ui_wellseis/src/qt/task_panel_base.cpp
#include <pwb/ui_wellseis/qt/task_panel_base.hpp>

#include <QLabel>
#include <QListWidget>
#include <QVBoxLayout>

#include <pwb/ui_wellseis/json_helpers.hpp>
#include <pwb/ui_wellseis/task_state.hpp>

namespace pwb::ui_wellseis::qt {

namespace {

QString qs(const std::string& text) {
    return QString::fromUtf8(text.data(), static_cast<int>(text.size()));
}

QLabel* add_value(QVBoxLayout* layout, const QString& label_text,
                  const QString& value_text, QWidget* parent) {
    auto* label = new QLabel(label_text, parent);
    label->setObjectName(QStringLiteral("WorkFieldLabel"));
    layout->addWidget(label);
    auto* value = new QLabel(value_text, parent);
    value->setObjectName(QStringLiteral("WorkFieldValue"));
    layout->addWidget(value);
    return value;
}

}  // namespace

TaskPanelBase::TaskPanelBase(QWidget* parent, const QString& object_name,
                             const QString& title, bool show_review_count)
    : QFrame(parent) {
    setObjectName(object_name);
    setMinimumWidth(200);

    auto* layout = new QVBoxLayout(this);
    layout->setContentsMargins(12, 12, 12, 12);
    layout->setSpacing(4);

    auto* title_label = new QLabel(title, this);
    title_label->setObjectName(QStringLiteral("MapDockTitle"));
    layout->addWidget(title_label);

    name_value_ = add_value(layout, QStringLiteral("当前任务"),
                            QStringLiteral("未选择预测任务"), this);
    adapter_value_ =
        add_value(layout, QStringLiteral("适配器"), QStringLiteral("—"), this);
    auto* status_label = new QLabel(QStringLiteral("状态"), this);
    status_label->setObjectName(QStringLiteral("WorkFieldLabel"));
    layout->addWidget(status_label);
    status_badge_ = new QLabel(QStringLiteral("待开始"), this);
    status_badge_->setObjectName(QStringLiteral("PwbBadge"));
    layout->addWidget(status_badge_);
    mean_probability_value_ = add_value(layout, QStringLiteral("平均概率"),
                                        QStringLiteral("—"), this);
    review_count_value_ =
        show_review_count
            ? add_value(layout, QStringLiteral("待复核区"),
                        QStringLiteral("0 个"), this)
            : nullptr;

    auto* list_label = new QLabel(QStringLiteral("任务列表"), this);
    list_label->setObjectName(QStringLiteral("WorkFieldLabel"));
    layout->addWidget(list_label);
    task_list_ = new QListWidget(this);
    task_list_->setObjectName(QStringLiteral("WorkListWidget"));
    connect(task_list_, &QListWidget::currentRowChanged, this,
            [this](int row) {
                if (!suppress_ && row >= 0) {
                    emit task_selected(row);
                }
            });
    layout->addWidget(task_list_, 1);
}

void TaskPanelBase::update_state(
    const std::vector<PredictionTaskSlice>& tasks,
    std::optional<int> selected_index) {
    tasks_ = tasks;
    const int active_row_core = task_panel_active_row(tasks, selected_index);
    const PredictionTaskSlice* task =
        active_row_core >= 0 ? &tasks[active_row_core] : nullptr;

    name_value_->setText(
        task != nullptr && !task->name.empty()
            ? qs(task->name)
            : QStringLiteral("未选择预测任务"));
    adapter_value_->setText(
        task != nullptr && !task->adapter_kind.empty()
            ? qs(task->adapter_kind)
            : QStringLiteral("—"));
    const TaskStatusToken token =
        task_status_token(task != nullptr ? task->status : "");
    status_badge_->setText(qs(token.label));
    status_badge_->setProperty("tone", qs(badge_tone_of(token.tone)));
    if (task != nullptr) {
        const std::string probability =
            json_str(task->probability_summary, "mean_probability", "");
        mean_probability_value_->setText(
            probability.empty() ? QStringLiteral("—") : qs(probability));
    } else {
        mean_probability_value_->setText(QStringLiteral("—"));
    }
    if (review_count_value_ != nullptr) {
        review_count_value_->setText(
            QString::number(task != nullptr ? task->review_area_count : 0) +
            QStringLiteral(" 个"));
    }

    // Keyed reconcile: rebuild rows but keep selection stable by task key —
    // the active task's key wins; else the previous selection's key.
    suppress_ = true;
    const QString previous_key =
        task_list_->currentItem() != nullptr
            ? task_list_->currentItem()->data(Qt::UserRole).toString()
            : QString();
    task_list_->clear();
    int active_row = -1;
    int previous_row = -1;
    const QString active_key =
        task != nullptr ? qs(task_key(*task)) : QString();
    for (std::size_t i = 0; i < tasks.size(); ++i) {
        const PredictionTaskSlice& item = tasks[i];
        const QString key = qs(task_key(item));
        auto* row_item = new QListWidgetItem(task_list_);
        row_item->setData(Qt::UserRole, key);
        row_item->setText(qs(task_row_text(item)));
        row_item->setToolTip(qs(task_row_tooltip(item)));
        if (!active_key.isEmpty() && key == active_key && active_row < 0) {
            active_row = static_cast<int>(i);
        }
        if (!previous_key.isEmpty() && key == previous_key &&
            previous_row < 0) {
            previous_row = static_cast<int>(i);
        }
    }
    const int row_to_select =
        active_row >= 0 ? active_row : previous_row;
    if (row_to_select >= 0) {
        task_list_->setCurrentRow(row_to_select);
    }
    suppress_ = false;
}

const std::vector<PredictionTaskSlice>& TaskPanelBase::tasks() const {
    return tasks_;
}

int TaskPanelBase::current_row() const {
    return task_list_->currentRow();
}

}  // namespace pwb::ui_wellseis::qt

File: legacy/python_reference/product/paleo_workbench/ui/pages/task_panel_base.py
from __future__ import annotations

from PySide6.QtCore import Signal
from PySide6.QtWidgets import (
    QFrame,
    QLabel,
    QListWidget,
    QListWidgetItem,
    QVBoxLayout,
)

from paleo_workbench.ui import tokens
from paleo_workbench.ui.components.badges import PwbBadge
from paleo_workbench.ui.modelview import reconcile_widget_items
from paleo_workbench.ui.workstation.state_language import (
    state_token,
    tone_to_badge,
)
from paleo_workbench.viz.prediction_helpers import active_prediction_task, field_value

#: 预测任务原始状态（英文字符串）→ state_language 任务词表键。
#: V11（01-ui-audit G3）：任务列表此前直接显示 pending/running 等英文
#: 裸串；统一经任务词表渲染（glyph + 中文 + tone）。
_TASK_STATUS_ALIASES = {
    "pending": "queued",
    "queued": "queued",
    "running": "running",
    "cancelling": "cancelling",
    "cancelled": "cancelled",
    "complete": "done",
    "completed": "done",
    "done": "done",
    "failed": "failed",
    "warning": "degraded",  # V11：warning=完成但有警示 → 降级完成（与 OperationRegistry/任务中心同义）
}


def task_status_token(status: str | None):
    """预测任务状态 → state_language 任务 StateToken（未知 → 排队中口径之外的
    「未知」不编造：回落 queued 之前先映射未知为 muted 的排队中）。"""
    key = _TASK_STATUS_ALIASES.get(str(status or "").strip().lower())
    if key is None:
        from paleo_workbench.ui.workstation.state_language import StateToken

        return StateToken("·", str(status or "待开始"), "muted")
    return state_token("task", key)


class TaskPanelBase(QFrame):
    """Shared left-hand task summary panel with selection list."""

    task_selected = Signal(int)

    def __init__(
        self,
        *,
        object_name: str,
        title: str,
        show_review_count: bool,
        parent=None,
    ):
        super().__init__(parent)
        self.setObjectName(object_name)
        # V9：固定 240 → 最小 200 地板。面板可在 splitter/dock 中自由
        # 调整宽度（页面对最大宽度的提升已解除上限）。
        self.setMinimumWidth(200)
        self._tasks: list = []
        self._suppress = False

        layout = QVBoxLayout(self)
        layout.setContentsMargins(
            tokens.PANEL_PADDING,
            tokens.PANEL_PADDING,
            tokens.PANEL_PADDING,
            tokens.PANEL_PADDING,
        )
        layout.setSpacing(tokens.SPACE_2)

        title_label = QLabel(title)
        title_label.setObjectName("MapDockTitle")
        layout.addWidget(title_label)

        self.name_value = self._add_value(layout, "当前任务", "未选择预测任务")
        self.adapter_value = self._add_value(layout, "适配器", "—")
        # V11：状态值经任务词表渲染为徽章（tone 随状态），不再是英文裸串。
        status_label = QLabel("状态")
        status_label.setObjectName("WorkFieldLabel")
        layout.addWidget(status_label)
        self.status_badge = PwbBadge("待开始", tone="neutral")
        self.status_value = self.status_badge  # 兼容既有属性名（setText 可用）
        layout.addWidget(self.status_badge)
        self.mean_probability_value = self._add_value(layout, "平均概率", "—")
        self.review_count_value = (
            self._add_value(layout, "待复核区", "0 个") if show_review_count else None
        )

        list_label = QLabel("任务列表")
        list_label.setObjectName("WorkFieldLabel")
        layout.addWidget(list_label)
        self.task_list = QListWidget()
        self.task_list.setObjectName("WorkListWidget")
        self.task_list.currentRowChanged.connect(self._on_row)
        layout.addWidget(self.task_list, 1)

    def _add_value(self, layout: QVBoxLayout, label_text: str, value_text: str) -> QLabel:
        label = QLabel(label_text)
        label.setObjectName("WorkFieldLabel")
        layout.addWidget(label)
        value = QLabel(value_text)
        value.setObjectName("WorkFieldValue")
        layout.addWidget(value)
        return value

    def _on_row(self, row: int) -> None:
        if not self._suppress and row >= 0:
            self.task_selected.emit(row)

    @staticmethod
    def _task_key(item) -> str:
        identity = field_value(item, "id", None)
        if identity:
            return f"id:{identity}"
        return f"name:{field_value(item, 'name', '') or ''}"

    def update_state(
        self,
        prediction_tasks: list | tuple | None,
        *,
        selected_index: int | None = None,
    ) -> None:
        tasks = list(prediction_tasks or [])
        self._tasks = tasks
        if selected_index is not None and 0 <= selected_index < len(tasks):
            task = tasks[selected_index]
        else:
            task = active_prediction_task(tasks)

        self.name_value.setText(field_value(task, "name", "") or "未选择预测任务")
        self.adapter_value.setText(field_value(task, "adapter_kind", "") or "—")
        token = task_status_token(field_value(task, "status", None) if task else None)
        self.status_badge.setText(token.label)
        self.status_badge.set_tone(tone_to_badge(token.tone))
        probability = (field_value(task, "probability_summary", {}) or {}).get(
            "mean_probability"
        )
        self.mean_probability_value.setText(
            str(probability) if probability is not None else "—"
        )
        if self.review_count_value is not None:
            self.review_count_value.setText(
                f"{len(field_value(task, 'review_areas', []) or [])} 个"
            )

        # V11：列表按键差分同步（同键任务不重建行，滚动/选择保持），
        # 行文本经任务词表渲染。键→任务一次映射（评审 P2：避免逐行全表
        # 键扫描的 O(n²)）。
        self._suppress = True
        keys = [self._task_key(item) for item in tasks]
        row_objects = dict(zip(keys, tasks))
        reconcile_widget_items(
            self.task_list,
            keys,
            make_item=lambda _key: QListWidgetItem(),
            update_item=lambda row_item, key: self._render_task_row(
                row_item, row_objects.get(key)
            ),
        )
        active_row = -1
        active_key = self._task_key(task) if task is not None else None
        for index, key in enumerate(keys):
            if active_key is not None and key == active_key:
                active_row = index
                break
        if active_row >= 0:
            self.task_list.setCurrentRow(active_row)
        self._suppress = False

    @staticmethod
    def _render_task_row(row_item, item) -> None:
        if item is None:
            return
        name = field_value(item, "name", "") or "未命名预测任务"
        token = task_status_token(field_value(item, "status", None))
        row_item.setText(f"{name} · {token.label}")
        row_item.setToolTip(f"{name}\n状态：{token.label}")
